### ISSUE TITLE
How to get output for a single image for VQA?

### ISSUE THREAD

**Issue by TheShadow29** [Cadene/block.bootstrap.pytorch#1]
Hi. I wanted to know if there is any easy way to get output for a single image (not necessarily a part of the val/test set)?

**Comment by Cadene** [comment]
Yes, you just create a batch of 1 image.

input=torch.randn(1,3,224,224)
vgg16.eval()
vgg16(input)

You can find this looking in the examples directory

**Comment by Cadene** [comment]
Actually, it is currently not possible, because the features extractor is not given.

If you have a pretrained features extractor (ConvNet or FasterRCNN), you could use it to extract the features, train a model and then it is easy.

[Here] you can find an example of a multimodal model (text encoder and image encoder) used to extract a unimodal embedding (so only the image part) from a random image.

**Comment by TheShadow29** [comment]
I have the features and the bounding box. The thing I am confused about is how to pass it to the model. As in what format should the input be.

**Comment by Cadene** [comment]
Awesome!

You can copy some parts of [this code](https://github.com/Cadene/recipe1m.bootstrap.pytorch/blob/master/recipe1m/api.py#L133)

Create a file (`block/run_single_image.py`):

```python
import torch
import bootstrap.lib.utils as utils
import bootstrap.engines as engines
import bootstrap.models as models
import bootstrap.datasets as datasets
from bootstrap.lib.logger import Logger
from bootstrap.lib.options import Options
from bootstrap.run import init_logs_options_files
from block.datasets.vqa_utils import tokenize_mcb

if __name__ == "__main__":
    Options(path_opts)
    utils.set_random_seed(Options()['misc']['seed'])
    init_logs_options_files(Options()['exp']['dir'], Options()['exp']['resume'])

    engine = engines.factory()
    engine.dataset = datasets.factory(engine)
    engine.model = models.factory(engine)
    engine.model.eval()
    engine.resume()

    img_emb_module = load_img_emb_module().cuda()
    
    # inputs
    img = torch.randn(1,3,224,244).cuda()
    question = 'What is the color of the white horse?'
    question_words = tokenize_mcb(question)
    question_wids = [engine.dataset['eval'].word_to_wid[word] for word in question_words]

    item = {
        'question': torch.LongTensor(question_wids),
        'lengths': torch.LongTensor([len(question_wids)]),
    }
    batch = engine.dataset['eval'].items_tf()([item])

    with torch.no_grad():
        batch['visual'] = img_emb_module(img)
        out = engine.model(batch)
        out = engine.model.network.process_answers(out)

    Logger()(out['answers'][0])
```

Execute it like this:
```
$ python -m block.run_single_image \
-o logs/vqa2/block/options.yaml \
--exp.resume best_eval_epoch.accuracy_top1 \
--dataset.train_split \
--dataset.eval_split val \
--misc.logs_name run_single_image
```

**Comment by TheShadow29** [comment]
Sweet. This works, with the minor change of unsqueezing `question` and `lengths` at index 0 and setting `batch = item` directly.

**Comment by Cadene** [comment]
Yeap you need to load the dataset, because I didn't save the word dictionaries in the state dict of the network.

**Comment by Karol-G** [comment]
Hi,

I also want to get the output for single images, but I don't quite understand the posted code.
@TheShadow29 or @Cadene I would be glad if one of you could post a working snipped.
If that is not possible it would be cool if you could explain this line of code to me at least:
`img_emb_module = load_img_emb_module().cuda()`
I use the features of the VQA2 dataset as I am solely working on this dataset.
